javascript/ql/src/Declarations/UnreachableMethodOverloads.qhelp

```diff
@@ -0,0 +1,64 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+    <p>
+        The TypeScript compiler has to choose which specific overload is called
+        when a method with multiple overloads is called.
+        The TypeScript compiler will always choose the textually first overload 
+        that does not give rise to any type errors with the arguments provided
+        at the function call.
+    </p>
+    <p>
+       This behavior can be unintuitive for programmers unfamiliar with the
+       type system in TypeScript, and can in some instances lead to situations
+       where a programmer writes an overloaded method where only the first 
+       overload can ever be used. 
+    </p>
+</overview>
+<recommendation>
+    <p>
+       Either reorder the method overloads if an overload with more type
+       parameters is placed before a similar overload with fewer parameters.  
+       Alternatively, collapse multiple overloads with identical parameter types by 
+       creating a single overload that returns a union of the return types 
+       from the multiple overloads. 
+    </p>
+</recommendation>
+<example>
+    <p>
+        In the below example a programmer has tried to express that a method 
+        can return multiple possible values by creating multiple overloads 
+        with identical parameter types. However, only the first overload 
+        will ever be selected by the TypeScript compiler.  
+    </p>
+    <sample src="examples/UnreachableMethodOverloads.ts" />
+    <p>
+       The error is fixed by merging the overloads into a single method 
+       signature that returns a union of the previous return types.
+    </p>
+    <sample src="examples/UnreachableMethodOverloadsGood.ts" />
+    
+    <p>
+        In the below example an interface <code>Foo</code> declares a method
+        <code>create()</code> with two overloads. The only difference between
+        the two overloads is the type parameter <code>T</code> in the first 
+        overload. The TypeScript compiler will always use the first overload 
+        when <code>create()</code> is called, as a default type will be used 
+        for the type parameter <code>T</code> if none is provided. 
+        This default type is <code>unknown</code> in TypeScript 3.5+, and 
+        <code>{}</code> in earlier versions. 
+    </p>
+    <sample src="examples/UnreachableMethodOverloadsTypeParameters.ts" />
+    <p>
+        In the below the error has been fixed by switching the order of the two
+        overloads. In this fixed version, if the <code>create()</code> method
+        is called with an explicit type argument the second overload will be
+        used, as the first overload would give rise to a type error.
+    </p>
+    <sample src="examples/UnreachableMethodOverloadsTypeParametersGood.ts" />
+</example>
+<references>
```

Comment:
No references?

Reply:
The only references I can come up with is a [pull request I made a while ago fixing an instance of the error](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32821/files), and the [TypeScript specification on how overloaded methods are resolved](https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#4.15.1).   
And I don't feel like any of those are really appropriate.

Reply:
Ok, point taken. Perhaps some of your team can make a decision on whether we should keep a References section at all (as it's empty). I don't recall being in this situation before but I'm sure this must occasionally happen.

Reply:
Hmmm. I looked through the qhelp of a bunch of other queries, and I changed my mind.   
I'll include the link to the TypeScript spec.